quiz.py

```diff
@@ -0,0 +1,144 @@
+
+        # 1. grab a key from the element_list randomly
+        correct_key = random.choice(self.elements_dict.keys())
+
+        # 2. Create a subset of the option_list for the quiz
+        correct_element = self.elements_dict[correct_key]
+        correct_option = self.options_dict[correct_key]
+        other_options = filter(lambda x: x != correct_option, set(self.options_dict.values()))
+        aux_list = random.sample(other_options, self.num_elements_quiz - 1)
+        quiz_options_set = set(aux_list)
+        quiz_options_set.add(correct_option)
+
+        # 3. Shuffle the new sub_option_list
+        quiz_options_list = list(quiz_options_set)
+        random.shuffle(quiz_options_list)
+        question = MultipleChoiceQuestion({correct_key: correct_element}, correct_option, quiz_options_list)
+
+        self.current_question = question
+        return question
+
+    def process_answer(self, answer):
+        correct = answer == self.current_question.correct_option.id
+
+        # update score
+        self.update_score(correct)
+
+        # 0. pop element from element list
+        self.elements_dict.pop(self.current_question.question_dict.keys()[0])
+
+        if not correct:
+            # save popped element in the seen_elements list
+            self.seen_elements_dict.update(self.current_question.question_dict)
+
+
+class PainterQuiz(ImageQuiz):
+    def __init__(self, num_elements_quiz=4):
+        super(PainterQuiz, self).__init__(num_elements_quiz=num_elements_quiz)
+
+    def initialize_image_quiz(self, num_elements=5):
+        stmt = exists().where(Painter.id == Painting.painter_id)
+        selected_painters = Painter.query.filter(stmt).order_by(func.random()).limit(num_elements).all()
+
+        painters_dict = dict()
+        paintings_dict = dict()
+        selected_paintings = []
+        for key, painter in enumerate(selected_painters):
+            paintings_aux = Painting.query.filter(Painting.painter_id == painter.id).order_by(func.random()).limit(
+                1).first()
+            selected_paintings.extend([paintings_aux])
+            painters_dict[key] = painter
+            paintings_dict[key] = paintings_aux  # get the first
+        return paintings_dict, painters_dict
+
+
+class SaintQuiz(ImageQuiz):
+    def __init__(self, num_elements_quiz=4):
+        # Elements are Paintings and Options are saints
+        super(SaintQuiz, self).__init__(num_elements_quiz=num_elements_quiz)
+
+    def initialize_image_quiz(self, num_elements=5):
+        selected_saints = Saint.query.filter(Saint.paintings.any()).order_by(func.random()).limit(num_elements).all()
+
+        saints_dict = dict()
+        paintings_dict = dict()
+        selected_paintings = []
+        for key, saint in enumerate(selected_saints):
+            paintings_aux = Painting.query.filter(Painting.saints.any(id=saint.id)).order_by(func.random()).limit(
+                1).first()
+            selected_paintings.extend([paintings_aux])
```

Comment:
I wanted to make it general enough so that maybe a couple or more paintings can be taken for each saint. Should I change it?